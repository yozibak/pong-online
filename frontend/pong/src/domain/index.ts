import { store } from "../data"
import { PlayerNumber, Position, Score } from "../data/types"
import { StateSnapshot } from "./resolvers"

export type RenderingState = {
  bars: Record<PlayerNumber, Position>
  ball: Position
  score: Score
}

export const getRenderingState = (): RenderingState => ({
  bars: {
    1: store.current.bars[1].position,
    2: store.current.bars[2].position,
  },
  ball: store.current.ball.position,
  score: store.current.score
})

export const getNetworkPayload = (): StateSnapshot => ({
  ball: store.current.ball,
  bars: store.current.bars
})
