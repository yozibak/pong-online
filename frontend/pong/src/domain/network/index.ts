import { GameID } from "../../config"
import { store } from "../../data"
import { PlayerCommand, Position } from "../../data/types"

export type NetworkPayload = {
  gameID: string
  playerNumber: number
  ball: {
    position: Position,
    movement: Position
  },
  bar: {
    position: Position,
    command: PlayerCommand
  }
  frameCount: number
}

const getNetworkPayload = (): NetworkPayload => {
  const playerNumber = store.current.playerNumber
  const frameCount = p.frameCount
  return {
    gameID: GameID,
    playerNumber,
    ball: {
      position: store.current.ball.position.position,
      movement: store.current.ball.movement.position
    },
    bar: {
      position: store.current.bars[playerNumber].position,
      command: store.current.bars[playerNumber].command
    },
    frameCount
  }
}
