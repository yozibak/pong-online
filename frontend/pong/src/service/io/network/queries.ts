export const sendData = /* GraphQL */ `
  mutation MyMutation(
    $ball: BallInput!
    $bar: BarInput!
    $frameCount: Int!
    $gameID: String!
    $playerNumber: Int!
  ) {
    sendData(
      bar: $bar
      frameCount: $frameCount
      gameID: $gameID
      playerNumber: $playerNumber
      ball: $ball
    ) {
      ball {
        missed
        movement {
          x
          y
        }
        position {
          x
          y
        }
      }
      bar {
        command
        position {
          x
          y
        }
      }
      frameCount
      gameID
      playerNumber
    }
  }
`

export const subscribeGame = /* GraphQL */ `
  subscription MySubscription($gameID: String!, $playerNumber: Int!) {
    subscribeGame(gameID: $gameID, playerNumber: $playerNumber) {
      ball {
        missed
        movement {
          x
          y
        }
        position {
          x
          y
        }
      }
      bar {
        command
        position {
          x
          y
        }
      }
      gameID
      playerNumber
      frameCount
    }
  }
`

export const sendHandShake = /* GraphQL */ `
  mutation MyMutation($body: String!, $gameID: String!, $playerNumber: Int!) {
    sendHandShake(body: $body, gameID: $gameID, playerNumber: $playerNumber) {
      body
      gameID
      playerNumber
    }
  }
`

export const subscribeHandShake = /* GraphQL */ `
  subscription MySubscription($gameID: String!, $playerNumber: Int!) {
    subscribeHandShake(gameID: $gameID, playerNumber: $playerNumber) {
      body
      gameID
      playerNumber
    }
  }
`
