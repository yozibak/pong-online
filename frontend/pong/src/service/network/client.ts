import { Amplify } from 'aws-amplify'
import { generateClient } from 'aws-amplify/api'

Amplify.configure({
  API: {
    GraphQL: {
      endpoint: process.env.VITE_API_ENDPOINT!,
      region: 'ap-northeast-1',
      defaultAuthMode: 'apiKey',
      apiKey: process.env.VITE_API_KEY,
    },
  },
})

export const client = generateClient()
