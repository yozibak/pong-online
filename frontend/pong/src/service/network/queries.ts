export const sendData = /* GraphQL */ `
  mutation MyMutation(
    $ball: BallInput!
    $bar: BarInput!
    $frameCount: Int!
    $gameID: String!
    $playerNumber: Int!
  ) {
    sendData(
      bar: $bar
      frameCount: $frameCount
      gameID: $gameID
      playerNumber: $playerNumber
      ball: $ball
    ) {
      ball {
        missed
        movement {
          x
          y
        }
        position {
          x
          y
        }
      }
      bar {
        command
        position {
          x
          y
        }
      }
      frameCount
      gameID
      playerNumber
    }
  }
`

export const subscribeGame = /* GraphQL */ `
  subscription MySubscription($gameID: String!, $playerNumber: Int!) {
    subscribeGame(gameID: $gameID, playerNumber: $playerNumber) {
      ball {
        missed
        movement {
          x
          y
        }
        position {
          x
          y
        }
      }
      bar {
        command
        position {
          x
          y
        }
      }
      gameID
      playerNumber
      frameCount
    }
  }
`
