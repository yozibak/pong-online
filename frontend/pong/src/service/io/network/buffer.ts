import { startSubscription } from '.'
import { PlayerNumber } from '../../../data/types'
import { opponentPlayerNumber } from '../../../domain/input/helpers'
import { NetworkPayload } from '../../../domain/output'

export const makeNetwork = () => {
  let subscription: ReturnType<typeof startSubscription>
  let listenTo: PlayerNumber
  let gameID: string
  let subscriptionHandler: (d: NetworkPayload) => void = () => undefined
  return {
    init(playerNumber: PlayerNumber, id: string) {
      listenTo = opponentPlayerNumber(playerNumber)
      gameID = id
      subscription = startSubscription(gameID, listenTo, (d) => subscriptionHandler(d))
    },
    updateHandler(handler: (d: NetworkPayload) => void) {
      subscriptionHandler = handler
    },
    stop() {
      subscription.unsubscribe()
    },
  }
}
