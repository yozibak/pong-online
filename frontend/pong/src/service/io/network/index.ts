import type { GraphQLSubscription } from 'aws-amplify/api'
import { PlayerNumber } from '../../../data/types'
import { NetworkPayload } from '../../../domain/output'
import { client } from './client'
import { sendData, subscribeGame } from './queries'

export const startSubscription = (
  gameID: string,
  playerNumber: PlayerNumber,
  onData: (payload: NetworkPayload) => void
) => {
  const subscription = client
    .graphql<GraphQLSubscription<{ subscribeGame: NetworkPayload }>>({
      query: subscribeGame,
      variables: {
        gameID,
        playerNumber,
      },
    })
    .subscribe({
      next: (value) => onData(value.data.subscribeGame),
      // eslint-disable-next-line no-console
      error: (e) => console.warn(e),
    })
  return subscription
}

export const sendDataToServer = async (payload: NetworkPayload) => {
  try {
    await client.graphql({
      query: sendData,
      variables: payload,
    })
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn(e)
  }
}
