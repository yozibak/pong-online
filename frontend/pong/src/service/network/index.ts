import type { GraphQLSubscription } from 'aws-amplify/api'
import { client } from './client'
import { PlayerNumber } from '../../data/types'
import { sendData, subscribeGame } from './queries'
import { NetworkPayload } from '../../domain/network'

export const startSubscription = (
  gameID: string,
  playerNumber: PlayerNumber,
  onData: (payload: NetworkPayload) => void
) => {
  const subscription = client
    .graphql<GraphQLSubscription<{ subscribeGame: NetworkPayload }>>({
      query: subscribeGame,
      variables: {
        gameID,
        playerNumber,
      },
    })
    .subscribe({
      next: (value) => onData(value.data.subscribeGame),
      // eslint-disable-next-line no-console
      error: (e) => console.warn(e),
    })
  return subscription
}

export const sendDataToServer = async (payload: NetworkPayload) => {
  try {
    await client.graphql({
      query: sendData,
      variables: payload
    })
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn(e)
  }
}
