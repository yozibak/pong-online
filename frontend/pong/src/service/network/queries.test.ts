import { NetworkPayload } from '../../domain/network'
import { sendData } from './queries'

test(`query construction`, () => {
  const payload: NetworkPayload = {
    gameID: 'gameID',
    playerNumber: 1,
    ball: {
      position: {
        x: 300,
        y: 300,
      },
      movement: {
        x: 10,
        y: 10,
      },
    },
    bar: {
      position: {
        x: 20,
        y: 300,
      },
      command: 'up',
    },
    frameCount: 2303,
  }
  const sendDataQuery = JSON.stringify({ mutation: sendData, variables: payload })
  expect(sendDataQuery).toMatchSnapshot()
})
