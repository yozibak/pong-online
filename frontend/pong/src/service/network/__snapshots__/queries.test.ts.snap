// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query construction 1`] = `"{"mutation":"\\n  mutation MyMutation(\\n    $ball: BallInput\\n    $bar: BarInput\\n    $frameCount: Int\\n    $gameID: String\\n    $playerNumber: Int\\n  ) {\\n    sendData(\\n      bar: $bar\\n      frameCount: $frameCount\\n      gameID: $gameID\\n      playerNumber: $playerNumber\\n      ball: $ball\\n    ) {\\n      ball {\\n        missed\\n        movement {\\n          x\\n          y\\n        }\\n        position {\\n          x\\n          y\\n        }\\n      }\\n      bar {\\n        command\\n        position {\\n          x\\n          y\\n        }\\n      }\\n      frameCount\\n      gameID\\n      playerNumber\\n    }\\n  }\\n","variables":{"gameID":"gameID","playerNumber":1,"ball":{"position":{"x":300,"y":300},"movement":{"x":10,"y":10}},"bar":{"position":{"x":20,"y":300},"command":"up"},"frameCount":2303}}"`;
